lib/ansible/modules/system/launchd.py

```diff
@@ -0,0 +1,236 @@
+        except OSError:
+            continue
+
+        for filename in files:
+            if filename == '%s.plist' % service_name:
+                return os.path.join(path, filename)
+
+
+def main():
+    # init
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(
+                required=True,
+                aliases=['service']
+            ),
+            state=dict(
+                choices=['started', 'stopped', 'restarted', 'reloaded'],
+            ),
+            enabled=dict(type='bool'),
+            sleep=dict(type='int',
+                       default=1),
+        ),
+        supports_check_mode=True,
+        required_one_of=[
+            ['state', 'enabled']
+        ],
+    )
+
+    launch = module.get_bin_path('launchctl', True)
+    service = module.params['name']
+    action = module.params['state']
+    rc = 0
+    out = err = ''
+    result = {
+        'name': service,
+        'changed': False,
+        'status': {},
+    }
+
+    rc, out, err = module.run_command('%s list' % launch)
+    if rc != 0:
+        module.fail_json(msg='running %s list failed' % launch)
+
+    running = False
+    found = False
+
+    for line in out.splitlines():
+        if line.strip():
+            pid, last_exit_code, label = line.split('\t')
+            if label.strip() == service:
+                found = True
+                if pid != '-':
+                    running = True
+                    result['status']['previous_state'] = 'running'
+                    result['status']['previous_pid'] = pid
+                else:
+                    result['status']['previous_state'] = 'stopped'
+                    result['status']['previous_pid'] = '-'
+                break
+
+    if not found:
+        module.fail_json(msg="Unable to find the service %s among active services." % service)
+
+    # Enable/disable service startup at boot if requested
+    if module.params['enabled'] is not None:
+        plist_file = find_service_plist(service)
+        if plist_file is None:
+            msg = 'Unable to infer the path of %s service plist file' % service
+            if not found:
+                msg += ' and it was not found among active services'
+            module.fail_json(msg=msg)
+
+        # Launchctl does not expose functionality to set the RunAtLoad
+        # attribute of a job definition. So we parse and modify the job definition
+        # plist file directly for this purpose.
+        service_plist = plistlib.readPlist(plist_file)
+        enabled = service_plist.get('RunAtLoad', False)
+
+        if module.params['enabled'] != enabled:
+            if not module.check_mode:
+                service_plist['RunAtLoad'] = module.params['enabled']
+                plistlib.writePlist(service_plist, plist_file)
+            result['changed'] = True
+
+    do = None
+    if action is not None:
+        if action == 'started' and not running:
+            do = 'start'
```

Comment:
In case a module was never loaded (`launchctl load <path-to-plist>`) than start/stop does not work.

I have replaced start/stop with load/unload and it works way better.

Reply:
Actually this is not entirely true. In fact if a service is stopped it needs to be started since load won't bring it up again.